odoo/addons/base/tests/test_expression.py

```diff
@@ -767,16 +768,211 @@ def test_30_normalize_domain(self):
+        any_domain = Domain('foo', 'any', Domain('bar', '=', 'baz'))
+        self.assertEqual(Domain(any_domain_list), any_domain)
+
+    def test_31_backwards_compatible_domain(self):
+        domain_a1 = Domain('a', '=', 1)
+        self.assertEqual(
+            list(domain_a1),
+            [('a', '=', 1)],
+            "Turn a domain into a list",
+        )
+        self.assertEqual(
+            domain_a1 + [('b', '=', 2)],
+            [('a', '=', 1), ('b', '=', 2)],
+            "Concatenation turns the domain into a list for backward compatibility",
+        )
+        self.assertEqual(
+            [('b', '=', 2)] + domain_a1,
+            [('b', '=', 2), ('a', '=', 1)],
+            "Concatenation turns the domain into a list for backward compatibility",
+        )
+        self.assertEqual(
+            ["!"] + domain_a1,
+            ~domain_a1,
+            "Special case that inverses the domain",
+        )
+
+        def normalize_domain(d):
+            return list(Domain(d))
+
+        self.assertEqual(
+            normalize_domain([('a', '=', 1), ('b', '=', 2)]),
+            ['&', ('a', '=', 1), ('b', '=', 2)],
+        )
+        self.assertEqual(
+            normalize_domain(['|', ('a', '=', 1), ('b', '=', 2)]),
+            ['|', ('a', '=', 1), ('b', '=', 2)],
+        )
+        self.assertEqual(
+            normalize_domain(['|', ('a', '=', 1), ('b', '=', 2), ('c', '=', 3)]),
+            ['&', '|', ('a', '=', 1), ('b', '=', 2), ('c', '=', 3)],
+        )
+        self.assertEqual(
+            normalize_domain([('a', '=', 1), '|', ('b', '=', 2), ('c', '=', 3)]),
+            ['&', ('a', '=', 1), '|', ('b', '=', 2), ('c', '=', 3)],
+        )
+        domain = [
+            ('a', '=', 1),
+            ('b.z', '=', 2),
+            '|', '|', ('c', '=', 3), ('d', '=', 4), ('e', '=', 5),
+            ('f', '=', 6),
+        ]
+        self.assertEqual(normalize_domain(domain), ['&', '&', '&'] + domain)
+
+        with self.assertRaises(ValueError):
+            normalize_domain(['&'])
+
+        with self.assertRaises(ValueError):
+            normalize_domain(['&', ('a', '=', 1)])
+
+        with self.assertRaises(ValueError):
+            normalize_domain([('a', '=', 1), '&', ('b', '=', 2)])
+
+        with self.assertRaises(ValueError):
+            normalize_domain([('a', '=', 1), '!'])
+
+        # rewrite rules when making a list
+        self.assertEqual(
+            list(Domain('foo', 'any', Domain('bar', '=', 'baz'))),
+            [('foo.bar', '=', 'baz')],
+            "'any' with a single domain condition is written with a dot notation",
+        )
+        self.assertEqual(
+            list(Domain('foo', 'any', Domain('bar', '=', 'baz') & domain_a1)),
+            [('foo', 'any', ['&', ('bar', '=', 'baz'), *domain_a1])],
+        )
+        self.assertEqual(
+            list(Domain('foo', 'in', [5])),
+            [('foo', '=', 5)],
+            "'in' with a single value represented as '='",
+        )
```

Comment:
I think that this is a mistake.  This transformation should be part of an optimization IMO.  Otherwise, what you see (`('foo', 'in', [5])`) is not what you get (`('foo', '=', 5)`).

IMO the domain construction should not modify conditions.  I don't have a definitive opinion on logic operators AND, OR and NOT, which may potentially simplify the domain right away, like in `Domain.FALSE & ... & ...` giving `Domain.FALSE`.

Maybe the transformation of `foo IN (5)` into `foo = 5` should be done just before converting it to SQL.  Otherwise this adds unnecessary burden on search methods.

Reply:
In `DomainCondition.new` some optimizations are already applied, there are comments which ones inside that function. Since we already check here the consistency of the domain (exclude/deprecate some usages), we can already optimize some domains and avoid re-creating a node during the optimization process - similarly to what is done for binary operators.

Both domains are semantically equivalent. `=` will be rewritten/optimized because it is not one of `STANDARD_CONDITION_OPERATORS`.